redis/redis.go

```diff
@@ -139,6 +139,56 @@ func (c *RedisStore[V]) SetEx(ctx context.Context, key string, value V, ttl time
 	return nil
 }
 
+func (c *RedisStore[V]) GetEx(ctx context.Context, key string) (V, time.Duration, bool, error) {
+	var out V
+	var ttl time.Duration
+
+	_, err := c.client.TxPipelined(ctx, func(pipe redis.Pipeliner) error {
+		getVal := pipe.Get(ctx, key)
+		getTTL := pipe.TTL(ctx, key)
+
+		_, err := pipe.Exec(ctx)
+		if err != nil && !errors.Is(err, redis.Nil) {
+			return fmt.Errorf("exec: %w", err)
+		}
+
+		if errors.Is(err, redis.Nil) {
+			return nil
+		}
+
+		ttl, err = getTTL.Result()
+		if err != nil {
+			return fmt.Errorf("TTL command failed: %w", err)
+		}
+
+		if ttl == -1 {
+			return fmt.Errorf("key %s does not have ttl set", key)
+		}
+
+		data, err := getVal.Bytes()
+		if data == nil {
+			return fmt.Errorf("get bytes: %w", err)
+		}
+
+		out, err = deserialize[V](data)
+		if err != nil {
+			return fmt.Errorf("deserialize: %w", err)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return out, ttl, false, fmt.Errorf("GetEx: %w", err)
+	}
+
+	if ttl == 0 {
```

Comment:
can this ever happen?

Reply:
I've changed the logic a little bit so it's more readable